scripts/staging/fm.dml

```diff
@@ -0,0 +1,91 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+# 
+#   http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+/**
+ * Factorization Machines.
+ */
+
+forward = function(matrix[double] X, matrix[double] w0, matrix[double] W, matrix[double] V)
+    return (matrix[double] y) {
+  /*
+   * Computes the model y. 
+   *
+   * Reference: 
+   *  - Factorization Machines, Steffen Rendle.
+   *
+   * Inputs:
+   *  - X : n examples with d features, of shape (n, d).
+   *  - w0: the global bias, of shape (1,).
+   *  - W : the strength of each feature, of shape (d, 1).
+   *  - V : factorized interaction terms, of shape (d, k).
+   *  
+   * Outputs:
+   *  - y : target vector, of shape (n, 1)
+   */
+   
+   y = w0 + X %*% W + 0.5 * rowSums( (X%*% V)^2 - (X^2 %*% V^2) )
+}
+   
+   
+// TODO : These gradients, from the paper, are the gradients w.r.t. the output y of the  
+// FM model, i.e., dy/dPARAM. We will want to optimize some loss function depending 
+// on the problem, so we'll want to compute dloss/dPARAM. 
+
+backward = function(matrix[double] dloss/dy, matrix[double] X, matrix[double] w0, matrix[double] W, matrix[double] V)
```

Comment:
shorten `dloss/dy` to `dy`, and then rename `dy` to `dout` if you also change `y` to `out` in the forward pass.